### PR TITLE
Add <Text> shadow support

### PR DESCRIPTION
**Pull request by Kudo**
Add three new TextStylePropTypes for \<Text>
- textShadowOffset
- textShadowRadius
- textShadowColor

### CODE REVIEW

**Comment by Kudo** [comment]
BTW, there is already a shadowOffset property for <Text> at https://github.com/facebook/react-native/blob/master/Libraries/Text/RCTShadowText.m#L367 but seems not being used.

I am trying to align the naming as CSS text-shadow and guessing the original shadowOffset is for \<View>, that's why the name is textShadowOffset

**Comment by satya164** [comment]
@Kudo Awesome. Can you give a screenshot? :)

**Comment by Kudo** [comment]
@satya164 Screenshots from UIExplorer

**Comment by nicklockwood** [comment]
This looks great! Thanks for doing the Android version too, that's amazing.

Good catch on the shadowOffset prop as well - not sure what that's doing there as it should be on RCTViewManaget, not RCTTextManager.

I'll review and merge this when I'm back in the office. @mkonicek can you check over the Android code?

**Comment by nicklockwood** [comment]
@kudo from the screenshots, I think there may be a small bug. The offset is roughly double in the Android screenshot to the iOS distance.

My guess is that the units on iOS may be in pixels, so you need to multiply by the `view.contentScaleFactor` to get consistent, display-independent offset on iOS.

**Comment by satya164** [comment]
This is so awesome. Thanks for working on this @Kudo . And extra thanks for adding Android support as well :D

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ShadowStyleSpan.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import android.text.style.CharacterStyle;
+
+public class ShadowStyleSpan extends CharacterStyle {
+  private final float mDx;
+  private final float mDy;
+  private final float mRadius;
+  private final int mColor;
+
+  public ShadowStyleSpan(
+    float dx,
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ShadowStyleSpan.java]: These all can fit in one line, as long as it's within 100 characters.

**Comment by Kudo** [comment]
@nicklockwood Thanks for pointing out for the small bug. The Android API shadowOffset unit is not well documented but it is px in reality.  6892d6b fixed the issue and I also align the UIExplorer sample between iOS and Android.

Here are updated screenshots, seems better!?

**Comment by Kudo** [comment]
@satya164 Good suggestion for the code style. 72b39be fixed this. Thank you.

**Comment by Kudo** [comment]
BTW to keep the change history during review and discussion, I will not squash commits in the mean time.
After passing review, I will do the squash.
Is it okay? Or let me know if you guys prefer squash first.
Thank you for reviewing.

**Comment by satya164** [comment]
@Kudo Doesn't really matter. Commits are squahsed by the bot during the import.

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Comment by Kudo** [comment]
So cool, thank you guys @satya164 @nicklockwood

**Comment by Kudo** [comment]
Thank you guys for getting this feature merged to upstream.

**Comment by nicklockwood** [comment]
@Kudo no problem, thanks for a very clean PR - makes our lives much easier :-)

**Comment by Kudo** [comment]
@nicklockwood I really like React Native, it's my pleasure to be able do something for RN. I will try to contribute more if possible. Thank you.

**Comment by avindra** [comment]
Saw this in the release notes. A bit confused, what's the difference between these properties and ```shadowColor```, ```shadowOpacity``` and friends?

**Comment by Kudo** [comment]
@avindra Did you mean <View>'s shadowColor and shadowOpacity?
Those are for View's rectangle shadow and the PR is about Text's drop shadow.
You might able to check [this playground](https://rnplay.org/apps/Rdz04Q) and differentiate between screenshots above.

**Comment by avindra** [comment]
Interesting.... I've been able to get text shadow working (on iOS, haven't tested Android) by setting ```shadowColor``` and ```shadowOpacity``` on a ```<Text>``` node with past releases. Guess that wasn't intentional. Will test new properties soon.

**Comment by Kudo** [comment]
@avindra The playground I posted is shadowColor/shadowOpacity on `<Text>` and the version is RN 0.13.0-rc or 0.11. Please trying to tap the "Tap to Play' button to check the result.
I think the playground is just as your description but the only difference is not sure which past releases of RN you tested before.

**Comment by avindra** [comment]
I've tested on 0.16 and 0.17. I do see that the yellow box shadow shows up, and not a text shadow, on rnplay. Hmmm....

**Comment by Kudo** [comment]
@avindra This is screenshot from my RN 0.17 iOS simulator, for your reference.

**Comment by kohver** [comment]
Hey guys, I have an error when do `textShadowOffset: {height: 1}` (without `width`).
I think it's wrong behavior.
image

**Comment by Kudo** [comment]
@kohver Sorry not handle this case well and #7119 fix the case.

**Comment by th3no0b** [comment]
I've just discovered that text shadows don't work with `{width: 0, height: 0}`, meaning shadows have to be slightly offset, is this intentional?

**Comment by ide** [comment]
@th3no0b that could be how the native platform implements shadows, a good way to answer your question is to write a sample with UIKit shadows / Android elevation.

**Pull request closed without merging**